[RFC] Port partial patches from vim

brcolow opened the pull request:
This PR ports a good chunk of the applicable patches from upstream vim dealing with partials. 

- [x] Allow partials to be used by jobs. (Thanks to @bfredl)
- [x] Allow partials to be used with dictwatchers. (Thanks to @bfredl)
- [x] Allow partials to be used with timers. (Thanks to @bfredl)
- [x] 7.4.1559
- [x] 7.4.1563
- [x] 7.4.1564
- [x] 7.4.1577
- [x] 7.4.1580
- [x] 7.4.1581
- [x] 7.4.1582
- [x] 7.4.1585
- [x] 7.4.1586
- [x] 7.4.1589
- [x] 7.4.1590
- [x] 7.4.1605
- [x] 7.4.1606
- [x] 7.4.1607
- [x] 7.4.1608
- [x] 7.4.1638
- [x] 7.4.1639
- [x] 7.4.1644 (Marked N/A)
- [x] 7.4.1645
- [x] 7.4.1656
- [x] 7.4.1715
- [x] 7.4.1718
- [x] 7.4.1719
- [x] 7.4.1720 (Marked as N/A)
- [x] 7.4.1731 (Marked as N/A, left note in vim_diff.txt)
- [x] 7.4.1836
- [x] 7.4.1839
- [x] 7.4.1842
- [x] 7.4.1860 (Was previously marked as N/A, but I took a test from it related to timers+ partials)
- [x] 7.4.1862 (Marked as N/A)
- [x] 7.4.1875

brcolow committed:
vim-patch:7.4.1559

Problem:    Passing cookie to a callback is clumsy.
Solution:   Change function() to take arguments and return a partial.

https://github.com/vim/vim/commit/1735bc988c546cc962c5f94792815b4d7cb79710

bfredl commented:
ref #5065 for list of follow-up patches. We also need to change timers/jobs etc to allow, either use vim representation or dedicated `callback_T` (ref #5022)

brcolow committed:
vim-patch:7.4.1563

Problem:    Partial test fails on windows.
Solution:   Return 1 or -1 from compare function.

https://github.com/vim/vim/commit/790500a8e65bee295ef51a59dfa67ecbaab8ea17

brcolow committed:
vim-patch:7.4.1564

Problem:    An empty list in function() causes an error.
Solution:   Handle an empty list like there is no list of arguments.

https://github.com/vim/vim/commit/346418c624f1bc7c04c98907134a2b284e6452dd

brcolow committed:
vim-patch:7.4.1577

Problem:    Cannot pass "dict.Myfunc" around as a partial.
Solution:   Create a partial when expected.

https://github.com/vim/vim/commit/ab1fa3955f25dfdb7e329c3bd76e175c93c8cb5e

brcolow committed:
vim-patch:7.4.1580

Problem:    Crash when using function reference. (Luchr)
Solution:   Set initial refcount. (Ken Takata)

https://github.com/vim/vim/commit/7a5c46a9df7ef01a4f6a620861c35400d5ad28d9

brcolow committed:
vim-patch:7.4.1581

Problem:    Using ":call dict.func()" where the function is a partial does
            not work.  Using "dict.func()" where the function does not take a
            Dictionary does not work.
Solution:   Handle partial properly in ":call". (Yasuhiro Matsumoto)

https://github.com/vim/vim/commit/65639032bb7b17996cd255d1508a1df4ad528a1f

brcolow committed:
vim-patch:7.4.1582

Problem:    Get E923 when using function(dict.func, [], dict). (Kent Sibilev)
            Storing a function with a dict in a variable drops the dict if the
            function is script-local.
Solution:   Translate the function name.  Use dict arg if present.

https://github.com/vim/vim/commit/6f2e4b36c9d9908e1cace2b1b96e2c154a837bc2

brcolow committed:
vim-patch:7.4.1585

Problem:    Partial is not recognized everywhere.
Solution:   Check for partial in trans_function_name(). (Yasuhiro Matsumoto)
            Add a test.

https://github.com/vim/vim/commit/d22a18928ebcb465393da1418bb88204b97badb1

brcolow committed:
vim-patch:7.4.1586

Problem:    Nesting partials doesn't work.
Solution:   Append arguments. (Ken Takata)

https://github.com/vim/vim/commit/8a1bb046378f4bc68d6a04af2eab80fb3ce04da6

brcolow committed:
vim-patch:7.4.1605

Problem:    Catching exception that won't be thrown.
Solution:   Remove try/catch.

https://github.com/vim/vim/commit/3905e291fe4375ca5c59efa9ffcb01a39c7be3a9

brcolow committed:
vim-patch:7.4.1589

Problem:    Combining dict and args with partial doesn't always work.
Solution:   Use the arguments from the partial.

https://github.com/vim/vim/commit/9e63f61cb01c70fd71652f54b2d01ee27b2a3534

brcolow committed:
vim-patch:7.4.1590

Problem:    Warning for shadowed variable. (Christian Brabandt)
Solution:   Move the variable into a local block.

https://github.com/vim/vim/commit/3f242a844e83a5a04943869f6e3bcbf8650dc465

brcolow committed:
vim-patch:7.4.1606

Problem:    Having type() handle a Funcref that is or isn't a partial
            differently causes problems for existing scripts.
Solution:   Make type() return the same value. (Thinca)

https://github.com/vim/vim/commit/953cc7fb139dc2ba8590f8b03a095b63f4e1208f

brcolow committed:
vim-patch:7.4.1607

Problem:    Comparing a function that exists on two dicts is not backwards
            compatible. (Thinca)
Solution:   Only compare the function, not what the partial adds.

https://github.com/vim/vim/commit/f0e86a0dbddc18568910e9e4aaae0cd88ca8087a

brcolow committed:
vim-patch:7.4.1638

Problem:    When binding a function to a dict the reference count is wrong.
Solution:   Decrement dict reference count, only reference the function when
            actually making a copy. (Ken Takata)

https://github.com/vim/vim/commit/e4eb6ff089e79e659acf33c17dd0fda7177de526

brcolow committed:
vim-patch:7.4.1639

Problem:    Invoking garbage collection may cause a double free.
Solution:   Don't free the dict in a partial when recursive is FALSE.

https://github.com/vim/vim/commit/5f436fcf9960c95702820d5ac1b8b612995f6c04

brcolow commented:
I am currently running into a consistent segfault at the same place in vim test suite:

https://travis-ci.org/neovim/neovim/jobs/171244522#L7968

If anyone can help with debugging, that would be very much appreciated.

brcolow committed:
vim-patch:7.4.1645

Problem:    When a dict contains a partial it can't be redefined as a
            function. (Nikolai Pavlov)
Solution:   Remove the partial when overwriting with a function.

https://github.com/vim/vim/commit/c5fbe8af4cd80789f831b78aa44ff0b238138769

brcolow committed:
vim-patch:7.4.1715

Problem:    Double free when a partial is in a cycle with a list or dict.
            (Nikolai Pavlov)
Solution:   Do not free a nested list or dict used by the partial.

https://github.com/vim/vim/commit/ddecc25947dbdd689d5bcaed32f298a08abdd497

brcolow committed:
vim-patch:7.4.1836

Problem:    When using a partial on a dictionary it always gets bound to that
            dictionary.
Solution:   Make a difference between binding a function to a dictionary
            explicitly or automatically.

https://github.com/vim/vim/commit/1d429610bf9e99a6252be8abbc910d6667e4d1da

brcolow committed:
vim-patch:7.4.1839

Problem:    Cannot get the items stored in a partial.
Solution:   Support using get() on a partial.

https://github.com/vim/vim/commit/2bbf8eff6fab16d86e7bcfc0da1962d31bec7891

brcolow committed:
vim-patch:7.4.1608

Problem:    string() doesn't handle a partial.
Solution:   Make a string from a partial.

https://github.com/vim/vim/commit/5c29154b521e9948190be653cfda666ecbb63b5b

brcolow committed:
vim-patch:7.4.1842

Problem:    get() works for Partial but not for Funcref.
Solution:   Accept Funcref.  Also return the function itself. (Nikolai Pavlov)

https://github.com/vim/vim/commit/03e19a04ac2ca55643663b97b6ab94043233dcbd

brcolow committed:
vim-patch:7.4.1875

Problem:    Comparing functions and partials doesn't work well.
Solution:   Add tests. (Nikolai Pavlov)  Compare the dict and arguments in the
            partial.

https://github.com/vim/vim/commit/8e759ba8651428995b338b66c615367259f79766

brcolow committed:
Fixes.

brcolow commented:
Update: There is still one failing assertion, but since it is preceded by an `unlet` I think it's related to the following issue:

Tests are passing, but we are leaking a lot of memory. This is because `partial_free` is not being called. Attempting to add `partial_unref(partial)` after line 18572 of `eval.c` causes segfaults (apparently a double free?) Any hints/help on this issue would be appreciated.

brcolow committed:
vim-patch:7.4.1719

Problem:    Leaking memory when there is a cycle involving a job and a
            partial.
Solution:   Add a copyID to job and channel.  Set references in items referred
            by them.  Go through all jobs and channels to find unreferenced
            items.  Also, decrement reference counts when garbage collecting.

https://github.com/vim/vim/commit/107e1eef1df3b786ad3ad49fbdb9e058649303b5

justinmk commented on src/nvim/eval/encode.c:
```diff
@@ -315,6 +315,55 @@ int encode_read_from_list(ListReaderState *const state, char *const buf,
+            *r++ = NUL; \
+          } \
+          ga_concat(gap, s); \
+          xfree(s); \
```
need to free `r` as well

jamessan replied:
Why? `s` exists to remember what to free, while `r` is being used to move through the allocated memory and populate it.

justinmk commented on src/nvim/eval/encode.c:
```diff
@@ -315,6 +315,55 @@ int encode_read_from_list(ListReaderState *const state, char *const buf,
+        ga_concat(gap, "function("); \
+        if (&pt->pt_name != NULL) { \
+          unsigned len; \
+          char_u *p, *r, *s; \
```
could move the declarations of `r` and `s` down to their first usage below

justinmk commented on src/nvim/eval/encode.c:
```diff
@@ -315,6 +315,55 @@ int encode_read_from_list(ListReaderState *const state, char *const buf,
+          } \
+          ga_append(gap, ']'); \
+        } \
+        if (pt != NULL && pt->pt_dict != NULL) { \
```
checking for NULL here (and line 347), but `&pt->pt_name` is used near the top.

justinmk commented on src/nvim/eval/encode.c:
```diff
@@ -315,6 +315,56 @@ int encode_read_from_list(ListReaderState *const state, char *const buf,
+            if (i > 0) { \
+              ga_concat(gap, ", "); \
+            } \
+            ga_concat(gap, encode_tv2string(&pt->pt_argv[i], NULL)); \
```
These calls to `encode_tv2string()` are at least part of the leak. `encode_tv2string()` allocates, then `ga_concat()` copies the result without freeing. Try this instead:

```c
    char *tofree = encode_tv2string(&pt->pt_argv[i], NULL);                                                                  
    ga_concat(gap, tofree);
    xfree(tofree);
```

Same for the other calls.

jamessan commented on src/nvim/version.c:
```diff
@@ -723,7 +723,6 @@ static int included_patches[] = {
   // 1723,
   // 1722 NA
   // 1721 NA
-  // 1720,
```
Usually that's done by adding `NA` after the patch number. :)

jamessan replied:
Oh, there was an extra 1720.

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -3481,6 +3559,13 @@ get({dict}, {key} [, {default}])
 		item is not available return {default}.  Return zero when
 		{default} is omitted.
+get({func}, {what})
+		Get an item with from Funcref {func}.  Possible values for
```
`Get item {what} from Funcref {func}`.

bfredl commented:
I'm looking into making them work for timers and jobs.

bfredl committed:
Make partials work with jobs, timers, and dictwatchers.

bfredl commented:
very WIP: https://github.com/bfredl/neovim/commit/100afd6f0f824bf562f009ce323fe6ee538d354d . I'll see if the design works well for jobs and dictwatchers also.

brcolow commented:
Very nice! Would you like me to wait to cherry pick these commits?

bfredl commented:
I hope to look into jobs tomorrow so maybe wait until then.

brcolow committed:
vim-patch:7.4.1860

Problem:    Using a partial for timer_start() may cause a crash.
Solution:   Set the copyID in timer objects. (Ozaki Kiichi)

https://github.com/vim/vim/commit/e3188e261569ae512fb1ae2653b57fdd9e259ca3

brcolow committed:
vim-patch:7.4.1731 Mark as NA

Leave a note in vim_diff.txt about it.

bfredl commented:
patch for dictwatchers here https://github.com/bfredl/neovim/commit/bdc220c37b7206e601c2edba85318d4d4ee8f9fe

two open questions:
- failed tests indicate a difference between vim and nvim when registering callbacks. in vim, registering a funcref to a function that doesn't exist is not an error, the error is first when trying to invoke the callback. In nvim this is an error already when defining the callback. I'm not sure which convention to adopt but it would be good to be internally consistent, at least.

- `dictwatcherdel()` relies on the identity of the callback to identify the watcher to remove. But this is a bit problematic for partials. When compared as viml values, partials are deeply compared by value, but this is not what we want, one could imagine two watchers being registered with the same function but with different self dicts, which for the moment `dictwatcherdel()` is called to remove one of them happens to be value-equal. For the moment `callback_equal` compares by reference to handle this, but it is not really good that it is inconsistent. Arguably, `dictwatcheradd()` should return a watcher id which can be sent to `dictwatcherdel()` as a single argument to remove the argument, in consistency with `timer_stop()`. (we should then deprecate but still support the old form only for funcrefs, for backwards compatibility). 

Jobs coming up next....

bfredl commented:
Yuck, `on_job_event` dispatches on the number of arguments that the function was _declared_ with. I wonder how many that know this and declares job callbacks with less than  three arguments, but I suppose we need to be backward compatible and still support this. Or if we want to remove this special case, the time is now...

justinmk commented:
@bfredl sounds like https://github.com/neovim/neovim/issues/3774.

bfredl commented:
Thanks for the reference. What do you think, would it be ok to break backwards-compatibility in this case? It might break a plugin or two or several, but I expect the error message to be easily understood and acted upon.

If not, we need to either add `callback_expected_arguments` (which I don't like, a callback is something you can _call_ and _free_ (+ mark for gc, 7.4.1860) , and _at worst_ compare for referential equallity, but nothing else) or add a "ignore surplus arguments" flag to call_func (which, special cases, we don't like either).

bfredl commented:
It is even worse , it seems we pass the implicit self argument even if it is not a dict function...

Example from job_spec:
```
    function! s:OnEvent(id, data, event)
      let userdata = get(self, 'user')
      call rpcnotify(g:channel, a:event, userdata, a:data)
    endfunction
```

should not work without marking the function as `dict`

bfredl commented:
very WIP, has all the mentioned problems and more (such as not gc-safe): https://github.com/bfredl/neovim/commit/72f320c0827eb0059b4f771bc52c86e6012b6e7e

bfredl commented:
I think this is what is needed to make callbacks gc safe https://github.com/bfredl/neovim/commit/a2dfa3c1e5eec8384d0e8904cbb335370977b481, but I'm not a gc expert, I might look into it more tomorrow.

@brcolow instead of cherrypicking, you can just reset your branch to that commit, unless you have more local commits not seen here. (in that case drop the two latest commits that is visible on github, they are replaced by my branch)

brcolow commented:
@bfredl Current failing/error tests:

```
./test/functional/helpers.lua:84: Vim(call):E118: Too many arguments for function: 1
stack traceback:
	./test/functional/helpers.lua:84: in function 'request'
	./test/functional/helpers.lua:142: in function 'nvim_command'
	./test/functional/helpers.lua:357: in function 'source'
	...is/build/neovim/neovim/test/functional/core/job_spec.lua:304: in function <...is/build/neovim/neovim/test/functional/core/job_spec.lua:303>
[  ERROR   ] ...is/build/neovim/neovim/test/functional/core/job_spec.lua @ 303: jobs jobwait will run callbacks while waiting (78.07 ms)

./test/functional/ui/screen.lua:297: Row 1 didn't match.
Expected:
|*~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|:function! g:JobHandler(job_id, data, event)         |
|:  :endfunction                                      |
|E127: Cannot redefine function JobHandler: It is in u|
|se                                                   |
|Press ENTER or type command to continue^              |
Actual:
|*^                                                     |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|~                                                    |
|                                                     |
To print the expect() call that would assert the current screen state, use
screen:snaphot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.  
stack traceback:
	./test/functional/ui/screen.lua:297: in function 'wait'
	./test/functional/ui/screen.lua:221: in function 'expect'
	...is/build/neovim/neovim/test/functional/core/job_spec.lua:272: in function <...is/build/neovim/neovim/test/functional/core/job_spec.lua:255>
[  ERROR   ] ...is/build/neovim/neovim/test/functional/core/job_spec.lua @ 255: jobs cannot redefine callbacks being used by a job (10515.25 ms)

...ovim/test/functional/ex_cmds/dict_notifications_spec.lua:233: Expected objects to be the same.
Passed in:
(number) 0
Expected:
(string) 'Vim(call):Function g:InvalidCb doesn't exist'
stack traceback:
	...ovim/test/functional/ex_cmds/dict_notifications_spec.lua:233: in function <...ovim/test/functional/ex_cmds/dict_notifications_spec.lua:232>
[  FAILED  ] ...ovim/test/functional/ex_cmds/dict_notifications_spec.lua @ 232: dictionary change notifications errors fails to add/remove if the callback doesn't exist (0.67 ms)

...ovim/test/functional/ex_cmds/dict_notifications_spec.lua:253: Expected objects to be the same.
Passed in:
(number) 0
Expected:
(string) 'Vim(function):E127: Cannot redefine function Watcher2: It is in use'
stack traceback:
	...ovim/test/functional/ex_cmds/dict_notifications_spec.lua:253: in function <...ovim/test/functional/ex_cmds/dict_notifications_spec.lua:246>
[  FAILED  ] ...ovim/test/functional/ex_cmds/dict_notifications_spec.lua @ 246: dictionary change notifications errors fails to replace a watcher function (1.29 ms)
```

bfredl commented:
Yes, it is the lack of the error that the function is not defined when the callback is created. I think the redefine error is basically the same question. VimL does not in general have references to the actual function (ufunc_T), only to its name. IMHO jobs (and dictwatchers) should neither, to avoid being a special case, and thus the tests should be updated. Instead we can add a test that an error is shown when the callback should called and it is not present.

justinmk commented:
Breaking back compatible is probably needed here. If there's an error and we can give clear scenario that plugin authors need to watch out for, should be ok.

bfredl commented:
The two latest commits on https://github.com/bfredl/neovim/tree/timerpartial are updating the tests, and cleaning up the argument count logic for jobs (#3774) respectively.

To be clear the real breaking changes are 

* job callbacks must take three arguments, not less.
* job callbacks must be defined as dict-functions to receive `self` (either by defining them as `func somedict.on_stdout(...)` or explicit `func s:callback(...) dict` )

The other changes are making errors into allowed actions, which I wouldn't expect to break anything (but you never know...)

brcolow commented:
Seems to be one failing test left:

```
[ RUN      ] ...d/neovim/neovim/test/functional/terminal/buffer_spec.lua @ 161: terminal buffer handles loss of focus gracefully
./test/functional/ui/screen.lua:297: Row 1 didn't match.
Expected:
|*ab^c                                               |
|{4:~                                                 }|
|{5:==========                                        }|
|rows: 2, cols: 50                                 |
|{2: }                                                 |
|{1:==========                                        }|
|                                                  |
Actual:
|*                                                  |
|                                                  |
|                                                  |
|                                                  |
|{8:E118: Too many arguments for function: SplitWindow}|
|                                                  |
|{UNEXPECTED foreground = 2:Press ENTER or type command to continue}           |
To print the expect() call that would assert the current screen state, use
screen:snaphot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.  
stack traceback:
    ./test/functional/ui/screen.lua:297: in function 'wait'
    ./test/functional/ui/screen.lua:221: in function 'expect'
    ...d/neovim/neovim/test/functional/terminal/buffer_spec.lua:182: in function <...d/neovim/neovim/test/functional/terminal/buffer_spec.lua:161>
[  ERROR   ] ...d/neovim/neovim/test/functional/terminal/buffer_spec.lua @ 161: terminal buffer handles loss of focus gracefully (10523.73 ms)
```

Which is triggered by the following:

```viml
    function! SplitWindow()
      new
      call feedkeys("iabc\<Esc>")
    endfunction

    startinsert
    call jobstart(['sh', '-c', 'exit'], {'on_exit': function("SplitWindow")})
```

bfredl commented:
Sorry, missed that one. Needs to add three arguments to the declaration of SplitWindow, like 
`function! SplitWindow(id, data, event)`

brcolow committed:
Add some tests and cleanup.

brcolow commented:
As far as I can tell this seems to be RDY.

brcolow commented on CMakeLists.txt:
```diff
@@ -418,7 +418,7 @@ if(NOT BUSTED_OUTPUT_TYPE)
     set(BUSTED_OUTPUT_TYPE "plainTerminal")
   else()
-    set(BUSTED_OUTPUT_TYPE "utfTerminal")
+    set(BUSTED_OUTPUT_TYPE "gtest")
```
Done.

bfredl commented:
Here is a test that seems to work:
```
describe('jobs with partials', function()
  local channel

  before_each(function()
    clear()
    channel = nvim('get_api_info')[1]
    nvim('set_var', 'channel', channel)
  end)

  it('works correctly', function()
    source([[
    function PrintArgs(a1, a2, id, data, event)
      call rpcnotify(g:channel, '1', a:a1,  a:a2, a:data, a:event)
    endfunction
    let Callback = function('PrintArgs', ["foo", "bar"])
    let g:job_opts = {'on_stdout': Callback}
    call jobstart(['echo'], g:job_opts)
    ]])
    eq({'notification', '1', {'foo', 'bar', {'', ''}, 'stdout'}}, next_msg())
  end)
end)
```

brcolow commented:
Seeing a failing test on Windows in `timer_spec.lua`:

```viml
 it('are triggered in blocking getchar() call', function()
    execute("call timer_start(50, 'MyHandler', {'repeat': -1})")
    nvim_async("command", "let g:c = getchar()")
    run(nil, nil, nil, 300)
    feed("c")
    local count = eval("g:val")
    ok(count >= 5) " <---- this condition
    eq(99, eval("g:c"))
  end)
```

bfredl commented:
Is it consistent (fails more often than not)? I guess it is timing issue, you could try change it to `count >= 4` .

brcolow commented:
Actually it seems it is not related to this PR at all, as I got the same failure on my PR for 7.1711:

https://ci.appveyor.com/project/neovim/neovim/build/9436/job/1c7e9suvxmq7l1rx#L6167

brcolow committed:
vim-patch:7.4.1718

Problem:    Coverity: not using return value of set_ref_in_item().
Solution:   Use the return value.

https://github.com/vim/vim/commit/d56374e25df0b317b01423a01f158157faa647fa

brcolow committed:
vim-patch:7.4.1720 Mark as NA

brcolow commented:
Rebased (fixed a merged conflict in `eval.c`).

jamessan commented:
> 7.4.1644 (Similar to 1608 but changes the implementation (apparently not the result) for how the partials are printed..but I don't understand why the implementation is moved from `tv2string` to `echo_string` )

In v7.4.1862, the guts of `tv2string` and `echo_string` were refactored into a common `echo_string_core`.  Is it worth integrating that into this PR?

jamessan reviewed (commented):
```vim
:let F = function('input', ['foo? '])
:quit
```

errors in nvim (`E951` and `E574`) because nvim is configured to store upper-case global variables in the shada file by default (`:help shada-!`).

In Vim, the variable isn't stored into the viminfo file, but the user also isn't annoyed by an error.

I think that needs to be fixed, especially since the user is _required_ to use an upper-case variable for a partial.

As far as 7.4.1644, it looks like we should either mark it NA or mark it as applied.  As far as I can tell, nvim's encoding already handles string vs. echo so the implementation you did when porting 7.4.1608 also handles 7.4.1644.

7.4.1862 looks like we should mark it NA (but I'm open to feedback on that).  I think that's covered under item 2 (either of them) in the "string() and echo behavior" section of `:help vim-features-changed`.

brcolow committed:
vim-patch:7.4.1644 Mark as NA

brcolow committed:
vim-patch:7.4.1862 Mark as NA

jamessan commented:
> 7.4.1862 looks like we should mark it NA (but I'm open to feedback on that).

Although, it may be worth picking out the relevant tests that were added in that patch.

jamessan commented:
TL;DR, LGTM.

> ```vim
> :let F = function('input', ['foo? '])
> :quit
> ```
> 
> errors in nvim (`E951` and `E574`) because nvim is configured to store upper-case global variables in the shada file by default (`:help shada-!`).
> 
> In Vim, the variable isn't stored into the viminfo file, but the user also isn't annoyed by an error.
> 
> I think that needs to be fixed, especially since the user is _required_ to use an upper-case variable for a partial.

Interestingly, there is a test (in test/functional/shada/variables_spec.lua) to verify that this scenario errors, while also continuing to write the rest of the data to the shada file.  In light of that, I guess I'm fine with the current behavior even if I don't think it's very useful to the user.  It can be addressed separately, if at all.

jamessan commented:
Thanks for all the work on this @brcolow and @bfredl!

justinmk commented on test/functional/core/job_spec.lua:
```diff
@@ -365,7 +368,7 @@ describe('jobs', function()
     end)
 
-    it('can be called recursively', function()
+    pending('can be called recursively', function()
```
so this is a regression?

brcolow commented:
🎉 🎈 🍰 😋 !

oni-link commented on src/nvim/eval/typval_encode.h:
```diff
@@ -248,6 +253,10 @@ static int name##_convert_one_value(firstargtype firstargname, \
       break; \
     } \
+    case VAR_PARTIAL: { \
+      TYPVAL_ENCODE_CONV_PARTIAL(tv->vval.v_partial); \
```
This function needs to do it's work iterative, but here it is possible to call `string_convert_one_value()` recursively, if the dictionary of the partial contains the partial:

```vim
func Callback() 
endf

let myDict={}
let Cb = function('Callback', [], myDict)
let myDict["1"]=Cb
echo Cb
```
```
=================================================================
==17185==ERROR: AddressSanitizer: stack-overflow on address 0x7fffbe461e5c (pc 0x7f17ad3888ae bp 0x7fffbe462300 sp 0x7fffbe461d80 T0)
    #0 0x7f17ad3888ad in _IO_vfprintf (/lib64/libc.so.6+0x468ad)
    #1 0x7f17ad38b660  (/lib64/libc.so.6+0x49660)
    #2 0x7f17ad388a54 in _IO_vfprintf (/lib64/libc.so.6+0x46a54)
    #3 0x4c65bf in __interceptor_vfprintf /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:1342
    #4 0x4c6692 in __interceptor_fprintf /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:1387
    #5 0x9b83d0 in string_convert_one_value /home/oni-link/git/neovim/src/nvim/eval/encode.c:443:1
    #6 0x996fcb in encode_vim_to_string /home/oni-link/git/neovim/src/nvim/eval/encode.c:443:1
    #7 0x991e4a in encode_tv2string /home/oni-link/git/neovim/src/nvim/eval/encode.c:811:3
    #8 0x9b8a54 in string_convert_one_value /home/oni-link/git/neovim/src/nvim/eval/encode.c:443:1
    #9 0x996fcb in encode_vim_to_string /home/oni-link/git/neovim/src/nvim/eval/encode.c:443:1
    #10 0x991e4a in encode_tv2string /home/oni-link/git/neovim/src/nvim/eval/encode.c:811:3
    #11 0x9b8a54 in string_convert_one_value /home/oni-link/git/neovim/src/nvim/eval/encode.c:443:1
    #12 0x996fcb in encode_vim_to_string /home/oni-link/git/neovim/src/nvim/eval/encode.c:443:1
    #13 0x991e4a in encode_tv2string /home/oni-link/git/neovim/src/nvim/eval/encode.c:811:3
    #14 0x9b8a54 in string_convert_one_value /home/oni-link/git/neovim/src/nvim/eval/encode.c:443:1
    .
    .
    .
```
So `TYPVAL_ENCODE_CONV_PARTIAL` should not make recursive calls to `string_convert_one_value()` by calling `encode_tv2string()`, instead it should push work on the stack like in the case for `VAR_DICT`.

haya14busa commented:
7.4.1644 is not N/A 

related: https://github.com/vim/vim/issues/708 haya14busa/incsearch.vim#125 https://github.com/haya14busa/incsearch.vim/issues/111

http://ftp.vim.org/vim/patches/7.4/7.4.1644

> Patch 7.4.1644
Problem:    Using string() on a partial that exists in the dictionary it binds
            results in an error. (Nikolai Pavlov)
Solution:   Make string() not fail on a recursively nested structure. (Ken
            Takta)
Files:      src/eval.c, src/testdir/test_partial.vim

justinmk commented:
It's"applied", you mean?

haya14busa commented:
```vim
let g:d = {}

" 3
let g:d.d = g:d
function g:d.test3()
endfunction

call string(g:d.test3)
" Error detected while processing /home/haya14busa/tmp/vim/test.vim:
" line    8:
" E724: unable to correctly dump variable with self-referencing container
```
(code borrowed from vim/vim#708)

It doesn't happen in Vim.

haya14busa commented:
yeah, I mean neovim should port patch 7.4.1644 as well, i think.

justinmk commented:
Agreed

mhinz commented:
Will be handled by: https://github.com/neovim/neovim/pull/5826

jamessan commented:
Isn't that being addressed in #5826?

ZyX-I commented:
There is self-referencing container, so error is expected. Behaviour diverted (error vs just `...`) at some point earlier, AFAIR Vim made `string()` not fail while it was initially failing while Neovim kept the behaviour present at the time of the fork and some time after.

Error is not related to funcrefs at all, #5826 only makes the code non-recursive and fixes bugs. Error when dumping something with reference cycles *is not a bug*.

justinmk commented:
@ZyX-I What should plugin authors do if they want to print a self-referencing container? (Without caring about roundtrips)

justinmk commented:
@ZyX-I In the [example](https://github.com/neovim/neovim/pull/5529#issuecomment-269512372) above, `string(g:d.test3)` only wants the function name. It shouldn't try to dump `g:d.d`, should it?

ZyX-I commented:
@justinmk There is `get()` for function names. For printing there is `:echo`, `string()` is for serializing. Though I think that `string()` errors out primary for historical reasons, do not remember that porting effects of Vim patches that allow `string()` to dump refcycles was discussed.

ZyX-I commented:
Also `silent! let repr = string(Var)` should work.

justinmk commented:
It does seem good to try to make `string()` as faithful as possible for roundtrips. That also makes it consistent with `msgpack_dump` and `json_encode` if I'm not mistaken.

> silent! let repr = string(Var) should work.

That works:

```
:silent! let foo = string(g:d.test3)                                                                                          
:echo foo " function('30', {'d': {E724@0}, 'test3': function('30')})
```

And the `{E724@0}` representation is much nicer than `...`. Perhaps we should always do that silently, and not raise an error? That avoids confusing plugin authors vs Vim, but keeps the nice granular error reporting.

ZyX-I commented:
@justinmk There is another variant: make something like e.g. `...@N` (literal three dots) or whatever actually mean “take N'th value from the stack”: it should be possible to patch `evalN` functions to put values they construct onto the stack so that such constructs are able to reference them, as well as it is possible (and is actually easier) to create anchors like in YAML.

ZyX-I commented:
E.g. lists `&l [*l]` (self-containing lists) and `[&l [], *l]` (list with two same lists as items) would be represented as

1. `[...@0]` and `[[], []]`: recursion survives and can be `eval()`ed back, but lists are no longer the same.
2. `@&l [@*l]` and `[@&l [], @*l]`: both recursion and identity survives and can be `eval()`ed back.
3. (should be the easiest to implement) `[@<l @=l]` and `[[@<l], @=l]`: 2, but unlike this marks are attached to `[` (in first case implementation is something like “save information that there is an anchor with its name, then when creating typval_T attach it to a named anchor in a dictionary and reset that information”, in second case it will be something like “when parsing a list after processing `[` check whether there is an anchor and, if yes, save newly created list”; in all cases mainly `eval7` is touched and `encode_vim_to_(string|echo)` additionally (not required, but they should be the first users)).

In all cases references are expression-local.

Note that all three systems are rather unique: I cannot remember any language which allows this.

ZyX-I commented:
(Of course, any programming language. Among markup languages at least YAML has this, not sure but very much likely there also is something on top of XML.)

haya14busa commented:
> Error when dumping something with reference cycles is not a bug.

 I agree. But, I mean it's inconsistent with Vim behavior.
 In addition, it's regression at least.

> @haya14busa What is the intention of the code in incsearch.vim that's trying to stringify a self-referencing container?

incsearch.vim doesn't try to stringify a self-referencing container, it tries to get a function name, not from dict but from funcref of dict function.
https://github.com/haya14busa/incsearch.vim/blob/213994fb3d75a30d3c1f8b965910f662d11cb1fd/autoload/incsearch/util.vim#L183

When vim introduced partial feature, string() of dict function changed to return binded dict and it can be recursive dict.
It introduced regression (e.g. incsearch.vim), and Vim fixed this behavior by not reporting error from string() of recursive dict.

I undestand incsearch.vim does hacky things and it can use `function(dict.Func, [arg], dict)` instead these days.
Maybe incsearch.vim should have change the code with version check, but since incsearch.vim supports old Vim and it works perfectly in Vim, I didn't fix it.

jamessan commented:
Just a few data points:

Partials were first added in [v7.4.1559] on 2016/03/14.  The string fix was added in [v7.4.1608] on 2016/03/19.  Using `get()` to access information about partials/funcrefs was added in [v7.4.1839]/[v7.4.1842] respectively on 2016/05/24.

Using [whohas] to see what versions of vim are packaged in the wild, only Ubuntu 16.04 (Xenial) has partials but not `get()` support, however it does have the `string()` fix.  However, Xenial happens to be an LTS release, so that will be around for awhile.

[v7.4.1559]: https://github.com/vim/vim/releases/tag/v7.4.1559
[v7.4.1608]: https://github.com/vim/vim/releases/tag/v7.4.1608
[v7.4.1839]: https://github.com/vim/vim/releases/tag/v7.4.1839
[v7.4.1842]: https://github.com/vim/vim/releases/tag/v7.4.1842
[whohas]: https://github.com/whohas/whohas

justinmk commented:
@ZyX-I If you plan to do any of the [approaches](https://github.com/neovim/neovim/pull/5529#issuecomment-271087495) you mentioned at some point in the future, it's probably best that we remove the error in the meantime. Because it won't be an error if/when one of those approaches is implemented.

ZyX-I commented:
@justinmk Third approach is generally the easiest to implement (first is also easy, but it would be harder to synchronize typval.c.h and evalN), but if you think any of them is a good idea then a separate issue needs to be opened for discussion.

Removing the error is a matter of moving

```C
#define TYPVAL_ENCODE_CONV_RECURSE(val, conv_type) \
    do { \
      char ebuf[NUMBUFLEN + 7]; \
      size_t backref = 0; \
      for (; backref < kv_size(*mpstack); backref++) { \
        const MPConvStackVal mpval = kv_A(*mpstack, backref); \
        if (mpval.type == conv_type) { \
          if (conv_type == kMPConvDict) { \
            if ((void *) mpval.data.d.dict == (void *) val) { \
              break; \
            } \
          } else if (conv_type == kMPConvList) { \
            if ((void *) mpval.data.l.list == (void *) val) { \
              break; \
            } \
          } \
        } \
      } \
      if (conv_type == kMPConvDict) { \
        vim_snprintf(ebuf, ARRAY_SIZE(ebuf), "{...@%zu}", backref); \
      } else { \
        vim_snprintf(ebuf, ARRAY_SIZE(ebuf), "[...@%zu]", backref); \
      } \
      ga_concat(gap, &ebuf[0]); \
      return OK; \
    } while (0)
```

up to replace previous occurrence of this macros, possibly also replacing `...` with `E724`, removing the `encode_vim_to_echo` (it will be identical to `encode_vim_to_string` then) and removing `did_echo_string_emsg` line from `encode_tv2string`. After this it is only needed to fix tests and possible travis failures. But I do not think I can do the last part (tests+CI) until weekend, though I may be wrong.
